Support having multiple crr locations configured in the queueProcessor

welansari opened the pull request:
QueueProcessor currently sets the destination host to the host of the last crr site in the bootstrap list, so when having multiple crr locations configured this will not work.

Issue: BB-662

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-662 contains:


* `9.0.5`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.5`

* `9.1.0`


Please check the `Fix Version/s` of BB-662, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

williamlardier reviewed (approved):
Changes LGTM (codecov still reports that all lines were not covered, yet the comment says the opposite...)

welansari commented:
The comments point to a previous run of codecov which is weird, but i added tests explicitly for those cases so they should be covered.

francoisferrand commented on extensions/replication/queueProcessor/QueueProcessor.js:
```diff
@@ -257,17 +258,16 @@ class QueueProcessor extends EventEmitter {
-                    if (dest.echo) {
-                        this._setupEcho();
-                    }
+            const dest = destConfig.bootstrapList.find(d => d.site === this.site);
```
for this I gather that the QueueProcessor only uses a single site from `dest.bootstrapList` : so should we not instead "select" the appropriate site's bootstrap list/endpoint(s) when building `destConfig`?

(since building every destConfig is done once, when loading the config, this would also allow to do it efficiently, without going searching linearly for every site)

welansari replied:
done

francoisferrand commented on extensions/replication/queueProcessor/task.js:
```diff
@@ -187,8 +187,6 @@ function initAndStart(zkClient) {
                 if (updatedSites.includes(site)) {
                     if (!destConfig[site]) {
```
is this condition still relevant : should we not update the destConfig in any case?

welansari replied:
We should not update the whole config in the other case as the QueueProcessor instance will loose reference to the config, but yes we should update the `replicationEndpoint` field. Added that.

francoisferrand commented on lib/Config.js:
```diff
@@ -313,7 +313,7 @@ class Config extends EventEmitter {
             transport: destConfig.sites?.[site]?.transport || destConfig.transport,
             auth: destConfig.sites?.[site]?.auth || destConfig.auth,
-            bootstrapList: this.bootstrapList,
+            replicationEndpoint: this.bootstrapList.find(endpoint => endpoint.site === site),
```
is there anything more useful in bootstrapList ?
* `servers` may be an array (or sometimes not? then that case should be handled)
* the `site` field is nto relvant once we build a specific sites destination endpoint

```suggestion
            endpoints: this.bootstrapList.find(endpoint => endpoint.site === site).servers,
```
or
```suggestion
            ...this.bootstrapList.find(endpoint => endpoint.site === site),
```

welansari replied:
There are other useful fields yes. For CRR: `servers`, `default`, `echo`. And for MultipleBackends: `type`

francoisferrand commented on extensions/replication/queueProcessor/QueueProcessor.js:
```diff
@@ -889,10 +884,8 @@ class QueueProcessor extends EventEmitter {
-                if (replicationEndpoint &&
-                replicationBackends.includes(replicationEndpoint.type)) {
+                if (this.destConfig.replicationEndpoint &&
+                replicationBackends.includes(this.destConfig.replicationEndpoint.type)) {
```
bad indent

francoisferrand commented on lib/Config.js:
```diff
@@ -313,7 +313,7 @@ class Config extends EventEmitter {
             transport: destConfig.sites?.[site]?.transport || destConfig.transport,
             auth: destConfig.sites?.[site]?.auth || destConfig.auth,
-            bootstrapList: this.bootstrapList,
+            replicationEndpoint: this.bootstrapList.find(endpoint => endpoint.site === site),
         };
```
probably not such a big deal today with limited number of locations, but could become one: we are looping through the whole list every time..... even though actually we are actually already looping over every site (looking at task.js)

```
        const bootstrapList = config.getBootstrapList();
        const siteNames = bootstrapList.map(i => i.site);

        // initialize per site destination configs
        const destConfig = {};
        siteNames.forEach(site => {
            destConfig[site] = config.getReplicationSiteDestConfig(site);
        });
```

probably not something we can change in the API, but when loading the config is could do some reprocessing to just linearly build just once both the list of sites and replicationEndpoints.

francoisferrand commented on extensions/replication/tasks/CopyLocationTask.js:
```diff
@@ -22,21 +22,14 @@ const MPU_GCP_MAX_PARTS = 1024;
 
 class CopyLocationTask extends BackbeatTask {
 
-    _getReplicationEndpointType() {
```
nit: to simplify the change, we could have kept this function and (greatly) simplify its implementation


```suggestion
    _getReplicationEndpointType() {
        return this.destConfig.replicationEndpoint.type
    }
```

francoisferrand commented on tests/functional/replication/pauseResumeState.js:
```diff
@@ -25,12 +25,16 @@ const repConfig = config.extensions.replication;
+    return {
+        auth: { type: 'skip', vault: 'skip' },
+        replicationEndpoint: repConfig.destination.bootstrapList
+            .find(endpoint => endpoint.site === site),
```
probably a simpler way to rework this: `firstSite` and `secondSite` are already build by indexing into the bootstrapList, not sure there is a point to add some lookup code (and possily bug) in the tests

Commit:
support multiple crr location in the queueProcessor

QueueProcessor currently sets the destination host to
the host of the last crr site in the bootstrap list.
This works in the 7.x branch as we filter the bootstrap
list to only contain the relevant site. This filtering
was done in the entrypoint of the QueueProcessor (i.e
task.js) which changed in 8.x and 9.x.

Issue: BB-662

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/9.1`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/9.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/9.0`



Please check the status of the associated issue BB-662.

Goodbye kerkesni.
